Switch to nbclassic and Python 3.12.11

danielhollas opened the pull request:
fixes #552 

We switch to nbclassic interface, which is now supported in appmode.

Notebook v7 is currently not supported in Appmode.

danielhollas committed:
WIP: Switch to nbclassic

danielhollas committed:
Python 3.11

danielhollas committed:
Fix integration test

danielhollas committed:
Merge branch 'main' into nbclassic

yakutovicha committed:
Python 3.11.14

yakutovicha committed:
Merge branch 'main' into nbclassic

yakutovicha committed:
Change jupyter base image.

yakutovicha committed:
Python 3.11.10

yakutovicha committed:
Select branch with migrated home app

yakutovicha committed:
Working version (tested locally)

yakutovicha committed:
Replace mamba run with conda run, otherwise hangs.

yakutovicha committed:
First working version

yakutovicha committed:
Merge branch 'main' into nbclassic

yakutovicha committed:
Merge branch 'main' into nbclassic

yakutovicha committed:
Bump appmode to 1.1.0

yakutovicha committed:
Python 3.12.11

yakutovicha committed:
Remove appmode enabling, it is enabled automatically.

yakutovicha committed:
Change aiidalab-home version (branch).

yakutovicha committed:
Fix logo and rotating gears.

edan-bainglass reviewed (commented):
Left comments

yakutovicha commented on stack/base/Dockerfile:
```diff
@@ -60,7 +60,7 @@ COPY pip.conf "${CONDA_DIR}/pip.conf"
 # https://github.com/aiidalab/aiidalab-docker-stack/issues/490
 # Install aiida-core and other shared requirements.
-RUN mamba update -y pip async_generator certipy && \
+RUN mamba update -y pip zstandard async_generator certipy && \
```
@danielhollas, what was the reason for this change?

yakutovicha replied:
not needed, in the end. So removed here: https://github.com/aiidalab/aiidalab-docker-stack/pull/479/commits/0fe29f209f943446d30e250b3f34ff5c6468a2be.

yakutovicha committed:
Trying to fix the tests

yakutovicha commented on stack/lab/Dockerfile:
```diff
@@ -32,8 +31,8 @@ RUN echo 'eval "$(_AIIDALAB_COMPLETE=bash_source aiidalab)"' >> "${CONDA_DIR}/et
-     git checkout v"${AIIDALAB_HOME_VERSION}" && \
-     pip install --no-user --quiet --no-cache-dir "./" && \
+     git checkout "${AIIDALAB_HOME_VERSION}" && \
+     pip install --no-user --no-cache-dir "./" && \
```
I used it for debugging, just to have more info in the output when installing the home app. Can revert.

yakutovicha committed:
Update Python version in tests: 3.11 -> 3.12

yakutovicha committed:
Fix container startup

- Replace mamba run with conda activate in startup scripts to fix
    hanging services (PostgreSQL and RabbitMQ now start properly)

yakutovicha committed:
DOCKER_STACKS_JUPYTER_CMD: notebook -> nbclassic

yakutovicha committed:
Work on healthchecks

yakutovicha committed:
Fix tests: use container-local paths and pip for git-installed packages

  - Use a container path for the venv fixture instead of host tmp_path,
    which is inaccessible from inside the Docker container.
  - Check aiidalab-home with pip show instead of mamba list, since it is
    installed from git via pip.
  - Handle branch names (e.g. "main") as aiidalab_home_version
  gracefully.

yakutovicha committed:
Incompatible packages: change ipywidgets==8.0.0 to 7.7.4

yakutovicha committed:
Fix conda envs_dirs: remove /opt/conda to prevent nesting error

Newer conda rejects creating environments directly inside another
conda environment. Remove /opt/conda from envs_dirs (the default
/opt/conda/envs is sufficient) and ensure ~/.conda/envs exists
with correct permissions for user-created environments.

yakutovicha commented:
I moved the parallel build of arm/amd architectures to another PR: https://github.com/aiidalab/aiidalab-docker-stack/pull/553, so we don't do everything here.

yakutovicha committed:
Set RMQ to 3.11.22

danielhollas commented on stack/base-with-services/before-notebook.d/20_start-postgresql.sh:
```diff
@@ -10,12 +10,14 @@ PSQL_LOGFILE="${PGDATA}/logfile"
 PSQL_START_CMD="pg_ctl --timeout=180 -w -l ${PSQL_LOGFILE} start"
 
-MAMBA_RUN="mamba run -n aiida-core-services"
+# Activate the aiida-core-services environment directly instead of using mamba run
```
I am very confused, do we know why `mamba run` suddenly doesn't work?

yakutovicha replied:
I don't know why, but my observation is that `mamba run` hangs when it calls a process that launches another process in the background. For some reason, it also waits for the background process to complete, which never happens. So the script never proceeds to completion.

~This, however, only happened for RQM; the change in PostgreSQL was only done so that RMA and PSQL launchers look similar~. [EDITED]

yakutovicha replied:
here is the investigation results: https://github.com/aiidalab/aiidalab-docker-stack/pull/479#issuecomment-4205169211

danielhollas replied:
Thanks for the detailed investigation! Great that you got to the bottom of it. 
I'll have a closer look but regardless of what solution we decide, I'd revert the changes to the start-postgresql script as they are not needed and I prefer the `mamba run` approach in general (running eval is not great and can have other side effects)

yakutovicha replied:
I re-ran the tests with `mamba run` for PSQL -- the problem is the same (pipe inheritance issue), so I remembered it wrongly. The changes are needed in both startup scripts.

I edited my comment above.

danielhollas commented on stack/base-with-services/before-notebook.d/30_start-rabbitmq.sh:
```diff
@@ -4,7 +4,14 @@ set -emx
+rabbitmq-server -detached
+
+# Wait a moment for RabbitMQ to start
+sleep 2
```
Is this really needed or was this for debugging? We spent a lot of time trying the make the container startup as fast as possible so would like to avoid sleeps unless really necessary.

yakutovicha replied:
I think you are right - it is not needed. None of the following startup scripts requires reassurance that RMQ is running. Deleted in https://github.com/aiidalab/aiidalab-docker-stack/pull/479/commits/d9609f9eb4ae5a9ec3541368881adc93f31f2d1a.

danielhollas commented on stack/docker-compose.base-with-services.yml:
```diff
@@ -1,17 +1,21 @@
         environment:
             TZ: Europe/Zurich
-            DOCKER_STACKS_JUPYTER_CMD: notebook
+            DOCKER_STACKS_JUPYTER_CMD: nbclassic
```
Let's remove this, I don't see a reason why this should be defined here, it's already defined in Dockerfile and should not really be changed.
```suggestion
```

danielhollas replied:
Done in #556

danielhollas commented on stack/docker-compose.base-with-services.yml:
```diff
@@ -1,17 +1,21 @@
             AIIDALAB_DEFAULT_APPS: ''
         volumes:
             - aiidalab-home-folder:/home/jovyan
+        healthcheck:
```
Can you explaing the rationale for adding this?

yakutovicha replied:
Yes. Before this was added (https://github.com/aiidalab/aiidalab-docker-stack/pull/479/commits/1a6fb5c86e0bc814238836e8688924b73fb3148c), all tests were failing because the standard health check was too restrictive: on GitHub runners, there was not enough time for everything to get started. As a result, the container was marked as unhealthy.

For example, all tests from the [previous commit](https://github.com/aiidalab/aiidalab-docker-stack/actions/runs/23460052563/job/68259365021) failed exactly for that reason.

danielhollas replied:
I guess the question is then why was this not needed before? Is the image starting slower now? (or is there something different?). Maybe we could try removing this now that the PR is basically done to see if this is still needed?

danielhollas commented on tests/test_base.py:
```diff
@@ -8,10 +8,11 @@
-    venv_path = tmp_path / ".venv"
-    aiidalab_exec(f"python -m venv {venv_path}")
-    return venv_path
+def venv(aiidalab_exec, nb_user):
```
Why this change? Previous code seems preferable and simpler, `tmp_path` fixture already takes care of the cleanup

yakutovicha replied:
Reverted in https://github.com/aiidalab/aiidalab-docker-stack/pull/479/commits/bd7992a9ba44985f8f05e797fe39deb408c7fa9f.

danielhollas reviewed (commented):
Some comments / questions

yakutovicha committed:
Merge branch 'main' into nbclassic

danielhollas commented on build.json:
```suggestion
      "default": "main"
```

danielhollas committed:
point aiidalab_home to main branch

danielhollas committed:
Merge branch 'main' into nbclassic

danielhollas commented:
I've merged in main to get aiida-core 2.8.0

danielhollas committed:
Fix test_correct_aiidalab_home_version_installed test

For some reason, mamba now lists this package
in canonicalize form as `aiidalab_home`

danielhollas committed:
Merge branch 'main' into nbclassic

yakutovicha committed:
Remove zstandard package

yakutovicha commented:
For the record, the failing tests are because we are constraining `ipywidgets~=8.0`, which means we won't be able to make the tests run before apps (QE app in this case) migrate to ipywidgets 8.x.

yakutovicha commented on stack/lab/Dockerfile:
```diff
@@ -80,7 +77,8 @@ USER ${NB_USER}
 
-RUN  mkdir -p /home/${NB_USER}/apps
+RUN mkdir -p /home/${NB_USER}/apps
+RUN mkdir -p /home/${NB_USER}/.tmp_notebooks
```
no:

<img width="903" height="46" alt="Image" src="https://github.com/user-attachments/assets/7863ed91-310a-4ba9-843b-069a05c26afc" />

I would then suggest removing that change and trying without it. If needed later, we can do a follow-up PR to not overload this one.

yakutovicha replied:
I checked the temporary notebooks created by appmode - they still live in the same folder as the source notebook. The name changed slightly, but that change has no consequence for us.

All in all, I see no need for a separate folder, as things are handled pretty much the same way as before.

I remoted the folder creation here: https://github.com/aiidalab/aiidalab-docker-stack/pull/479/commits/8a51479bf7ac18d5303853a076f9da978ea69e66

yakutovicha committed:
Merge branch 'main' into nbclassic

yakutovicha committed:
Remove pip install appmode

yakutovicha commented on stack/lab/Dockerfile:
```diff
@@ -44,12 +43,10 @@ RUN ipywidgets_major=$(mamba list -f ipywidgets | tail -1 | awk '{print $2}' | c
-# Enable appmode extension
-RUN jupyter nbextension enable --py --sys-prefix appmode && \
-     jupyter serverextension enable --py --sys-prefix appmode
+RUN pip install appmode==1.1.0 --no-cache-dir --no-user
```
Indeed, seems not needed anymore. Removed in https://github.com/aiidalab/aiidalab-docker-stack/pull/479/commits/060ed4a248780b1df808668694965aa556283aca.

yakutovicha committed:
Remove ~/.tmp_notebooks folder creation

yakutovicha committed:
Revert test_base.py to original state

yakutovicha committed:
start_rabbitmq.sh: removed sleep 2 and the outdated arm64 comment

yakutovicha commented:
@edan-bainglass, @danielhollas: Here is the result of the investigation regarding `mamba run` vs `conda activate`.

By default, conda run sets up pipes for stdout/stderr to capture the subprocess output. The flow is:

  1. `mamba run -n aiida-core-services pg_ctl start` spawns `pg_ctl` as a child process
  2. `pg_ctl`'s stdout/stderr are connected to pipes back to mamba run
  3. mamba run reads from those pipes until EOF, then exits

**What happens when starting the deamons**

When `pg_ctl` starts or `rabbitmq-server -detached` runs:

  1. The command forks a long-running daemon process.
  2. The daemon inherits the pipe file descriptors from its parent (stdout/stderr).
  3. The parent process exits normally.
  4. `mamba run` sees the parent exit, but it's still waiting for EOF on the pipe.
  5. EOF on a pipe only occurs when every process holding the write end closes it.
  6. The daemon still holds the pipe's write end open (inherited at fork).
  7. The daemon never exits -> pipe never gets EOF -> mamba run hangs forever

An alternative solution is suggested [here](https://github.com/conda/conda/issues/9412) is to use `--no-capture-output`

Here is the summary of all tries:


Approach | Result | Notes
-- | -- | --
mamba run (original) | HANGS | Pipe inheritance by forked daemons
mamba run --no-capture-output | BROKEN | Generates exec (in wrapper script -> exec): invalid option (mamba bug)
conda run --no-capture-output | WORKS | Avoids pipe capture without the exec bug
conda activate + direct commands | WORKS | To me, this is the cleanest way.

yakutovicha commented on stack/base-with-services/before-notebook.d/20_start-postgresql.sh:
```diff
@@ -10,12 +10,12 @@ PSQL_LOGFILE="${PGDATA}/logfile"
 PSQL_START_CMD="pg_ctl --timeout=180 -w -l ${PSQL_LOGFILE} start"
 
-MAMBA_RUN="mamba run -n aiida-core-services"
+CONDA_RUN="conda run -n aiida-core-services"
```
Here it is: https://github.com/aiidalab/aiidalab-docker-stack/pull/479#issuecomment-4205169211

yakutovicha commented:
@danielhollas, @edan-bainglass - to me, the PR is ready to go, please re-review.

danielhollas commented:
@yakutovicha thanks so much, I'll have a look (but will not have time today). Note that this is still blocked on the RabbitMQ upgrade, we need to resolve #510 first. But I agree that otherwise this looks ready.

Can you please merge main? That should remove some of the changes to the docker compose files.

danielhollas commented:
@yakutovicha I just launched the latest image from this PR and I am seeing a Notebook v7 warning banner:

<img width="2242" height="233" alt="image" src="https://github.com/user-attachments/assets/5c0c6cb7-d9b3-4aa3-b8bd-6511c0fcd247" />

I believe it was hidden in the earlier versions of this PR, perhaps that change got lost?

EDIT: Adding `"--NotebookApp.show_banner=False "\` to `NOTEBOOK_ARGS` should resolve this, see: https://github.com/aiidalab/aiidalab-docker-stack/pull/455/changes

edan-bainglass commented on build.json:
```diff
@@ -16,7 +16,7 @@
     },
     "AIIDALAB_HOME_TAG": {
-      "default": "v25.02.0"
+      "default": "main"
```
Why unpin?

danielhollas replied:
We should make a new release of aiidalab-home

edan-bainglass commented on docker-bake.hcl:
```diff
@@ -21,7 +21,7 @@ variable "AIIDALAB_HOME_TAG" {
 
 variable "JUPYTER_BASE_IMAGE" {
-  default = "jupyter/minimal-notebook:python-${PYTHON_VERSION}"
+  default = "quay.io/jupyter/minimal-notebook:python-${PYTHON_VERSION}"
```
Why?

danielhollas replied:
Newer version of jupyter stack image are no longer published on Dockerhub

edan-bainglass commented on tests/test_lab.py:
```diff
@@ -22,7 +22,7 @@ def test_correct_aiidalab_version_installed(package_info, aiidalab_version):
 
 def test_correct_aiidalab_home_version_installed(package_info, aiidalab_home_tag):
-    info = package_info("aiidalab-home")
+    info = package_info("aiidalab_home")
```
Was this a typo before?

danielhollas replied:
Newer mamba version in this image uses this canonicalized package name in its `mamba list` output

edan-bainglass reviewed (approved):
LGTM. Just a few comments, but good to go.

danielhollas committed:
Merge branch 'main' into nbclassic

danielhollas committed:
Temporarily Remove QeApp from integration tests

danielhollas committed:
Remove traitlets install in base image

We've been manually updating traitlets to 5.9 for
performance reasons, but the new python3.12 base
image contains newer traitlets version already.

danielhollas commented:
Pushed some small changes, hope that's okay. Disabled QeApp integration tests for now, but before this is merged it would be good to have at least a rough idea what it would take to upgrade QeApp for this image. 

May I ask, what kind of testing has been done on this image so far? I think at the very least we should test individual AWB widgets.

Marking this as blocked until #510 is merged (hope to make some progress on that over weekend) and we need to first release the 2.8 based image anyway.

danielhollas commented:
This is very minor, but I am seeing this warning when using pip in this image

<img width="2222" height="209" alt="image" src="https://github.com/user-attachments/assets/dcc86e53-5e91-4595-8863-18e4eba1c002" />

We can add `ENV PIP_USE_FEATURE="build-constraint"` in the base image to get rid of it (where we set PIP_CONSTRAINT).

danielhollas committed:
Merge branch 'main' into nbclassic

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

danielhollas commented:
> Marking this as blocked until https://github.com/aiidalab/aiidalab-docker-stack/pull/510 is merged (hope to make some progress on that over weekend) and we need to first release the 2.8 based image anyway.

The RMQ update has been merged and I've updated this PR here. The changes here look good to me, but I think we should release a new image with aiida-core 2.8.0 before merging this, see #564. I've also discussed with @yakutovicha that it would be good to do a bit more testing, in particular to have at least one app upgraded and confirmed working, so that we do not run into any further surprises.

Please also address my previous comments: (EDIT: I've done the first two) 
- [x] ~~https://github.com/aiidalab/aiidalab-docker-stack/pull/479#issuecomment-4227314870 (pip deprecation warning)~~ This turned out to be difficult so I've given up for now. This problem should resolve by itself in future pip versions
- [x] https://github.com/aiidalab/aiidalab-docker-stack/pull/479#issuecomment-4214098514 (hiding Notebook 7 banner)
- [ ] https://github.com/aiidalab/aiidalab-docker-stack/pull/479#discussion_r3061893545 (making a new release of aiidalab-home)

danielhollas committed:
Bump aiidalab package to v26.05.2

danielhollas committed:
Include pre-releases in integration tests

danielhollas committed:
Hide Notebook v7 banner

danielhollas committed:
Get rid of pip deprecation warning

danielhollas committed:
Merge branch 'main' into nbclassic

danielhollas committed:
Try to fix venv test

danielhollas committed:
Revert "Try to fix venv test"

This reverts commit 01213f788c709921962402d7541b8c9bfce4a6db.

danielhollas committed:
Upgrade pip in venv first

danielhollas committed:
Revert "Upgrade pip in venv first"

This reverts commit 847d5b12c8ce8178dcb7a0f50c5d980ea802dc9b.

danielhollas committed:
Revert "Get rid of pip deprecation warning"

This reverts commit 594e31f8fb56e6da4bb14aca63ac5d9464354e20.

danielhollas committed:
Merge branch 'main' into nbclassic

danielhollas reviewed (commented):
I think once we make the `aiidalab-home` release we can merge this.

danielhollas committed:
Bump aiidalab-home tag

danielhollas committed:
Update aiidalab to 26.06.0
